Modify rule S836: Adapt to LaYC (#2596)
SonarSource/rspec: rules/S836/cfamily/rule.adoc

Variables should be initialized before their use to avoid unexpected
behaviors due to garbage values.

== Why is this an issue?

A local variable of the built-in type (like `int`, `float`, and also pointers),
declared without initial value, is not initialized to any particular value.
Consequently, code that uses such a variable without previously assigned to

> them does not have defined behavior. For example:

Author: Good suggestion.

  return *p; // Noncompliant: p have garbage value
}
----

Similarly, structures that aggregate such variables, such as arrays or `struct`/`class`
types without a constructor, will not initialize thier members, when declared without
an initializer:

[source,cpp]
----
struct Aggregate {
  int i;
  float f;
};

void aggregates() {
  int* intArray[5]; // each element of array is not initializer
  Aggregate aggr; // members aggr.i, agrr.f are not initialized
  Aggregate aggrArray[2]; // members of each element are not initialied
}
----

Finally, allocatign objects of such builtin or aggregates types on heap, also does not
initialize their values:

[source,c]
----
void usingMalloc() {
  int* intArr = (int*)malloc(sizeof(int) * 10); // each of 10 allocated integers is not initialized
}
----

This also applies when `new` is used in `C++`:

[source,cpp]
----
void usingNew() {
  Aggregate* aggrPtr = new Aggregate; // members of allocated Aggregate are not initialized
  Aggregate* aggrArr = new Aggregate[5]; // members of each of 5 Aggregate objects are not initialized
}
----

=== What is the potential impact?

In the best-case scenario, using garbage value will cause the

> program to behave non-deterministically at runtime.

Author: If you ship it user, crashing it worse. But I understand this depends, and should not be in rule.

> program to behave non-deterministically at runtime.

Author: I just removed it and have:
> Using garbage values will cause the program to behave
non-deterministically at runtime.

The program may produce a different output or crash depending
on the run.

In some situations, loading a variable may expose a sensitive
data, like a password, that was previously stored in the same
location. Leading to vulnerability, where such defect

> is used as a gadget for extracting information from the instance

Author: I believe that in vulnerability reasearch world, the exploit word is avoided. And they refers to gadget. I will change it to "that uses this defect" if that is ok with you/

void uses() {
  static int staticInt;
  return globInt     // Compliant: all zero initialized
       + globTab[2]
       + globAggr.f
       + globPart.x
       + staticInt;
}
----

== How to fix it

Commonly, the use of an uninitialized object is an indication of a defect in code,
where either variable initialization was skipped on some code paths, or
the object is used by mistake. Generally, such problems can be addressed by:

* Initializing variable on the declaration with valid value
* Assigning to the variable on the code path(s) that was missing initialization

Whenever possible, it is preferable to initialize a variable with its final
value on the declaration, as this eliminates possiblity of this defect occuring.

=== Code examples

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
int function(int flag, int b) {
  int a;
  if (flag) {
    a = b;
  }
  return a; // Noncompliant: "a" has not been initialized in all paths
}
----

==== Compliant solution

Initializing variable in all code paths:

[source,c,diff-id=1,diff-type=compliant]
----
int function(int flag, int b) {
  int a;
  if (flag) {
    a = b;
  } else {
    a = 10;
  }

>   return a; // Complaint

Reviewer: ```suggestion
  return a; // Compliant
```
Please find&replace it in other places too

}
----

Skipping path that leads to read of uninitialized value:

[source,c,diff-id=1,diff-type=compliant]
----
int function(int flag, int b) {
  int a;
  if (flag) {
    a = b;
  } else {
    return 10;
  }
  return a; // Complaint
}
----

Providing valid initial value:

> [source,cpp]

Reviewer: Why did you change the language to `cpp`?

Author: Accident, I was having all examples marked as `cpp` in first place.

----
int function(int flag, int b) {
  int a = 10;
  if (flag) {
    a = b;
  }
  return a; // Complaint
}
----

Initializing value in the defintion:

[source,cpp]
----
int function(int flag, int b) {
  int const a = flag ? b : 10;
  return a; // Complaint
}
----

=== Going the extra mile

With the addition of lambda in {cpp}11, it is possible to initialize
variable on declaration without creating a separate function
to compute value:

[source,cpp]
----
int x = [&] { // capture all context by reference
  if (someCondition())
    return computeVar1();
  /* peform more computations */
  return other;

> }(); // invoke lambda just (immediately)  after creation

Reviewer: ```suggestion
}(); // invoke the lambda immediately (right after creation)
```

----

Such pattern is referred as _Immediately invoked function expression_
(IIFE) or _Imediately invoked lambda_.
Furthermore, with the addition of structured binding in {cpp}17, it is
possible to declare multiple variables whose values are coupled:

[source,cpp]
----

> auto [px, py. pz] = [&] {

Reviewer: ```suggestion
auto [px, py, pz] = [&] {
```

  if (x_dir) {

>     return std::make_tuple(1.0, 0, 0);
>   } else if (y_dir) {
>     return std::make_tuple(0, 1.0, 0);
>   } else {
>     assert(z_dir);
>     return std::make_tuple(0, 0, 1.0);

Reviewer: ```suggestion
    return std::make_tuple(1, 0, 0);
  } else if (y_dir) {
    return std::make_tuple(0, 1, 0);
  } else {
    assert(z_dir);
    return std::make_tuple(0, 0, 1);
```
To make it compile, and avoid the hairy `0., 1.0, 0.`

  }
}();
----

== Resources

=== Documentation

- C++ reference - https://en.cppreference.com/w/cpp/language/storage_duration[Storage class specifiers]

=== External coding guidelines

> * https://cwe.mitre.org/data/definitions/457[MITRE, CWE-457] - Use of Uninitialized Variable

Reviewer: The link should be applied to the document, leaving the standard (source) as plain text

Author: Could you elaborate. I used that exact spelling in other rule.

Author: OK, I checked the link formatting doc, and I think got it correct now.

Reviewer: I think it should be:
source - https://link[document]
```
* MITRE - https://cwe.mitre.org/data/definitions/457[CWE-457 Use of Uninitialized Variable]
```
But I can be wrong

Author: I used the first form. Grep in RSPEC shows only the old forms.

Author: I have checked with Best Practices, and we should just CWE as source:
https://github.com/SonarSource/rspec/blob/master/docs/link_formatting.adoc

* MISRA C:2004, 9.1 - All automatic variables shall have been assigned a value before being used.
* MISRA {cpp}:2008, 8-5-1 - All variables shall have a defined value before they are used.

=== Related rules

> * S2107 detect fields being left uninitialized after invocation of constructor

Reviewer: ```suggestion
* S2107 detects fields being left uninitialized after the invocation of a constructor
```

  return *p; // Noncompliant: p has garbage value
}
----

Similarly, structures that simply aggregate variables of built-in types, such as arrays or `struct`/`class`
types without a constructor, will not initialize their members when declared without
an initializer:

[source,cpp]
----
struct Aggregate {
  int i;
  float f;
};

void aggregates() {
  int* intArray[5]; // each element of array is not initializer
  Aggregate aggr; // members aggr.i, agrr.f are not initialized
  Aggregate aggrArray[2]; // members of each element are not initialized
}
----

Finally, allocating objects of builtin or such aggregates types on the heap,
also does not initialize their values:

[source,c]
----
void usingMalloc() {
  int* intArr = (int*)malloc(sizeof(int) * 10); // each of 10 allocated integers is not initialized
}
----

This also applies when `new` is used in {cpp}:

[source,cpp]
----
void usingNew() {
  Aggregate* aggrPtr = new Aggregate; // members of allocated Aggregate are not initialized
  Aggregate* aggrArr = new Aggregate[5]; // members of each of 5 Aggregate objects are not initialized
}
----

=== What is the potential impact?

Using garbage values will cause the program to behave

> non-deterministically at runtime.

Reviewer: ```suggestion
nondeterministically at runtime.
```

The program may produce a different output or crash depending
on the run.

In some situations, loading a variable may expose a sensitive
data, such as a password that was previously stored in the same
location, leading to a vulnerability that uses such a defect
as a gadget for extracting information from the instance
of the program.

Finally, in {cpp}, outside of a few exceptions related to the uses
of `unsigned char` or `std::byte`, loading data from an uninitialized
variable causes undefined behavior. This means that the compiler
is not bound by the language standard anymore, and the program
has no meaning assigned to it. As a consequence, the impact
of such a defect is not limited to the use of garbage values.

=== Why is there an issue for a class with a default constructor?

In {cpp}, a class can define a default constructor invoked when an
object of the given type is created. Such a constructor is called even
if a variable is declared without any initializer.
However, if the constructor code omits the initialization
of a member that itself does not have the default constructor,

> the member will remain uninitialized. And reading

Reviewer: ```suggestion
the member will remain uninitialized  (See also S2107). And reading
```

from it will produce a garbage value:

[source,cpp]
----
struct Partial {
  // x is not initialized
  Partial() : y(10.0) {}

  int x;
  float y;
};

> void initialized() {

Reviewer: ```suggestion
int initialized() {
```

  Partial p; // constructor is called
  // or even Partial p{};

>   return p.x; // Non-complaint: reading an uninitialized variable

Reviewer: ```suggestion
  return p.x; // Non-compliant: reading an uninitialized variable
```

}
----

> === Exceptions: global, static, and thread-local variables.
> 
> All the variables with static storage duration, meaning: global, static,
> or thread-local variables, are zero-initialized before the initializer
> is evaluated. As a consequence, any such variable or member of such
> object has a defined value even if no initializer is specified.

Reviewer: I would lean towards the standard section title "Exceptions", and  clarify it inside the section:

```suggestion
=== Exceptions

This rule does not flag the variables with static storage duration, meaning:
global, static, and thread-local variables.

All the variables with static storage duration are zero-initialized before the initializer is evaluated.
As a consequence, any variable or member of such an object has a defined value even if no initializer is specified.
```

[source,cpp]
----
int globInt;
int globTab[10];
Aggregate globAggr;
Partial globPart; // x member is zero-initialized

> void uses() {

Reviewer: ```suggestion
int uses() {
```

  static int staticInt;
  return globInt     // Compliant: all zero-initialized
       + globTab[2]
       + globAggr.f
       + globPart.x
       + staticInt;
}
----

== How to fix it

Commonly, the use of an uninitialized object is an indication of a defect in the
code, where either variable initialization was skipped on some code paths, or

> the object is used by mistake. Generally, such problems can be addressed by:

Reviewer: This is quite subjective, so feel free to discard it.

I notice that you use quite a lot of passive voice.
In general, it hinders comprehension and leads to more complex sentences.
It is also against [our guidelines](https://github.com/SonarSource/rspec/blob/master/docs/tone_guide.adoc#guidelines).

In most places in this rule description, however, it makes sense. Indeed, it is clearer to say "constructor is called" or "variable is initialized", rather than "C++ abstract machine skips the constructor call", or "initializes a variable".

But here, I think you could improve the readability by using active voice:
```suggestion
the object is used by mistake. Generally, you can address such problems by:
```

Author: Good suggestion. Applied.

* Initializing variables on the declaration with a valid value
* Assigning to the variable on the code path(s) that was missing initialization

Whenever possible, it is preferable to initialize a variable with its final
value on the declaration, as this eliminates the possibility of this defect
occurring.

=== Code examples

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
int function(int flag, int b) {
  int a;
  if (flag) {
    a = b;
  }
  return a; // Noncompliant: "a" has not been initialized in all paths
}
----

==== Compliant solution

> Initializing variable in all code paths:

Reviewer: ```suggestion
Initializing variable on all code paths:
```

[source,c,diff-id=1,diff-type=compliant]
----
int function(int flag, int b) {
  int a;
  if (flag) {
    a = b;
  } else {
    a = 10;
  }
  return a; // Compliant
}
----

Skipping path that leads to the read of an uninitialized value:

[source,c]
----
int function(int flag, int b) {
  int a;
  if (flag) {
    a = b;
  } else {
    return 10;
  }

>   return a; // Complaint

Reviewer: ```suggestion
  return a; // Compliant
```

}
----

Providing a valid initial value:

[source,cpp]
----
int function(int flag, int b) {
  int a = 10;
  if (flag) {
    a = b;
  }

>   return a; // Complaint

Reviewer: ```suggestion
  return a; // Compliant
```

}
----

Initializing value in the definition:

[source,c]
----
int function(int flag, int b) {
  int const a = flag ? b : 10;

>   return a; // Complaint

Reviewer: ```suggestion
  return a; // Compliant
```

}
----

=== Pitfalls

Initializing the variable to zero at the declaration is not always the right
solution to fix the issue, as it may lead to logic errors if such a value

> is not handled correctly. For example, setting an `age` field of the `Employee`
> the structure may break assumptions of retirement handling code. Or more commonly,

Reviewer: ```suggestion
is not handled correctly. For example, setting an `age` field of an `Employee`
structure may break assumptions of retirement handling code. Or more commonly,
```

> setting a pointer to `NULL`, will turn dereference of an uninitialized value

Reviewer: ```suggestion
setting a pointer to `NULL` will turn dereference of an uninitialized value
```

> into a null pointer dereference.

Reviewer: With no `-` both adjectives apply to the "dereference", i.e., it is equivalent to "pointer null dereference"
```suggestion
into a null-pointer dereference.
```

> === Going the extra mile

Reviewer: I think it is worth mentioning that in this case you can often make your variable `const`. And also that using `const` variables makes you more disciplined about your initialization.

Author: Making variable `const` prevents moving them, or (if I recall correctly) prevents named return initialization. I agree that using const when possible is better, but I think it is not necessary if you want to avoid uninitialized variables.

With the addition of lambdas in {cpp}11, it is possible to initialize
variables on declaration without creating a separate function

> to the compute value:

Reviewer: ```suggestion
to compute the value:
```

[source,cpp]
----
int x = [&] { // capture all context by reference
  if (someCondition())
    return computeVar1();
  /* perform more computations */
  return other;
}(); // invoke the lambda immediately (right after creation)
----

Such pattern is referred to as an _Immediately invoked function expression_
(IIFE) or _Imediately invoked lambda_.
Furthermore, with the addition of structured binding in {cpp}17, it is
possible to declare multiple variables whose values are coupled:

[source,cpp]
----
auto [px, py, pz] = [&] {
  if (x_dir) {
    return std::make_tuple(1, 0, 0);
  } else if (y_dir) {
    return std::make_tuple(0, 1, 0);
  } else {
    assert(z_dir);
    return std::make_tuple(0, 0, 1);
  }
}();
----

== Resources

=== Documentation

> - C++ reference - https://en.cppreference.com/w/cpp/language/storage_duration[Storage class specifiers]

Reviewer: Not sure if this makes any difference. for consistency
```suggestion
* C++ reference - https://en.cppreference.com/w/cpp/language/storage_duration[Storage class specifiers]
```
